core/trino-main/src/main/java/io/trino/operator/exchange/UniformPartitionRebalancer.java

```diff
@@ -0,0 +1,360 @@
+ *      Writer 3 -> Partition 1 (100MB) + Partition 2 (100MB) + Partition 3 (100MB) ->  300 MB
+ * <p>
+ * After scaling:
+ *      Writer 1 -> Partition 1 (50MB) + Partition 3 (50MB) -> 100 MB
+ *      Writer 2 -> Partition 2 (50MB) -> 50 MB
+ *      Writer 3 -> Partition 1 (150MB) + Partition 2 (150MB) + Partition 3 (150MB) ->  450 MB
+ */
+@ThreadSafe
+public class UniformPartitionRebalancer
+{
+    private static final Logger log = Logger.get(UniformPartitionRebalancer.class);
+    private static final long MIN_REBALANCE_PHYSICAL_WRITTEN_BYTES = DataSize.of(100, MEGABYTE).toBytes();
+
+    // If the percentage difference between the two writers with maximum and minimum physical written bytes
+    // since last rebalance is above 0.7 (or 70%), then we consider them skewed.
+    private static final double SKEWNESS_THRESHOLD = 0.7;
+
+    private final List<Supplier<Long>> writerPhysicalWrittenBytesSuppliers;
+    private final long writerMinSize;
+    private final int numberOfWriters;
+
+    private final AtomicReference<List<List<Integer>>> partitionToWritersAssignment;
+    private final AtomicLongArray writerPhysicalWrittenBytesAtLastRebalance;
+
+    @GuardedBy("this")
+    private final Map<WriterPartitionId, Long> writerPartitionIdToRowCount = new HashMap<>();
+
+    @GuardedBy("this")
+    private final long[] writerWrittenBytesSinceLastRebalance;
+
+    @GuardedBy("this")
+    private final long[] writerRowCountSinceLastRebalance;
+
+    @GuardedBy("this")
+    private final long[] partitionWrittenBytesAtLastRebalance;
+
+    public UniformPartitionRebalancer(
+            List<Supplier<Long>> writerPhysicalWrittenBytesSuppliers,
+            int partitionCount,
+            int numberOfWriters,
+            long writerMinSize)
+    {
+        this.writerPhysicalWrittenBytesSuppliers = requireNonNull(writerPhysicalWrittenBytesSuppliers, "writerPhysicalWrittenBytesSuppliers is null");
+        this.writerMinSize = writerMinSize;
+        this.numberOfWriters = numberOfWriters;
+
+        this.writerPhysicalWrittenBytesAtLastRebalance = new AtomicLongArray(numberOfWriters);
+        this.writerWrittenBytesSinceLastRebalance = new long[numberOfWriters];
+        this.writerRowCountSinceLastRebalance = new long[numberOfWriters];
+        this.partitionWrittenBytesAtLastRebalance = new long[partitionCount];
+
+        List<List<Integer>> partitionToWritersAssignment = new ArrayList<>();
+        for (int i = 0; i < partitionCount; i++) {
+            partitionToWritersAssignment.add(new CopyOnWriteArrayList<>(ImmutableList.of(i % numberOfWriters)));
+        }
+        this.partitionToWritersAssignment = new AtomicReference<>(partitionToWritersAssignment);
+    }
+
+    public List<Integer> getWriterIds(int partitionId)
+    {
+        return partitionToWritersAssignment.get().get(partitionId);
+    }
+
+    public synchronized void addPartitionsRowCount(Map<WriterPartitionId, Integer> partitionsRowCount)
```

Comment:
If this is per page the I suggest to make it lock free (read write lock maybe). It's not really that atomic operation